File: web/lib/brief.ts
/**
 * Executive Brief generator (Layer 4).
 *
 * Deterministic first: it assembles a complete, board-ready brief purely from
 * the structured analytical output, so it works with every AI provider off. If
 * a provider IS configured, the AI layer is handed that same structured
 * evidence and asked to rewrite the narrative sections. The recommendations and
 * every number always come from the deterministic layer, never the model.
 */

import "server-only";
import {
  getKpis, getRanking, getRecommendations, getForecast, getScenarioBaseline,
  type Kpi, type Recommendation,
} from "./queries";
import { generate, isConfigured, providerLabel } from "./ai/provider";
import { PERSONAS } from "./ai/personas";

export type ExecutiveBrief = {
  provenance: { engine: "deterministic" | "ai-enhanced"; provider: string };
  cycle: string;
  summary: string;
  insights: string[];
  risks: string[];
  recommendations: { title: string; action: string; priority: string; domain: string }[];
  nextActions: string[];
  evidence: Record<string, unknown>;
};

const byKey = (kpis: Kpi[], key: string) => kpis.find((k) => k.metric_key === key);
const fmtDelta = (d: number | null, unit = "") =>
  d == null ? "" : `${d >= 0 ? "up" : "down"} ${Math.abs(d).toFixed(1)}${unit}`;

export async function buildExecutiveBrief(): Promise<ExecutiveBrief> {
  const [kpis, recs, skillPrem, region, channel, ctcForecast, topList] = await Promise.all([
    getKpis(),
    getRecommendations(),
    getRanking("skill_salary_premium"),
    getRanking("region_placement_rate"),
    getRanking("channel_effectiveness"),
    getForecast("median_ctc_cycle"),
    getRanking("most_demanded_skills"),
  ]);

  const cycle = kpis[0]?.context?.replace(/^(Cycle|FY)\s*/i, "") ?? "the latest period";
  const projected = ctcForecast.find((p) => p.is_forecast);
  const topSkill = skillPrem[0];
  const topItem = topList[0];
  const bestChannel = channel[0];
  const weakRegion = region[0];
  const moneyUnit = kpis.find((k) => k.unit === "LPA" || k.unit === "Cr")?.unit ?? "";

  const fmtKpi = (k?: Kpi | null) => {
    if (!k) return "";
    if (k.unit === "%") return `${k.value.toFixed(1)}%`;
    if (k.unit === "LPA") return `₹${k.value.toFixed(1)} LPA`;
    if (k.unit === "Cr") return `₹${k.value.toFixed(0)} Cr`;
    return k.value.toLocaleString("en-IN", { maximumFractionDigits: 0 });
  };

  // --- deterministic narrative (module-agnostic) ---------------------------
  const kpiSentence = kpis.slice(0, 3).map((k) =>
    `${k.label.toLowerCase()} at ${fmtKpi(k)}` +
    (k.delta != null ? ` (${fmtDelta(k.delta, k.unit === "%" ? " points" : k.unit ? ` ${k.unit}` : "")})` : "")
  ).join(", ");
  const summary =
    `For ${cycle}, the headline position is ${kpiSentence}. ` +
    (projected ? `The trajectory projects to ₹${projected.yhat.toFixed(0)} ${moneyUnit} next period.` : "");

  const insights = [
    topSkill && `${topSkill.entity} carries the clearest premium at ${topSkill.metric >= 0 ? "+" : ""}${topSkill.metric.toFixed(1)}%, marking where value concentrates.`,
    !topSkill && topItem && `${topItem.entity} is the single largest line in the book.`,
    bestChannel && `${bestChannel.entity} leads the channel mix at ${bestChannel.metric.toFixed(1)}%.`,
    weakRegion && `${weakRegion.entity} is the weakest region at ${weakRegion.metric.toFixed(1)}%, below every other region.`,
  ].filter(Boolean) as string[];

  const highRisks = recs.filter((r) => r.priority === "High");
  const risks = [
    ...highRisks.map((r) => r.observation),
    weakRegion && `${weakRegion.entity} runs materially below the other regions and concentrates downside risk.`,
  ].filter(Boolean).slice(0, 4) as string[];

  const recommendations = recs.slice(0, 5).map((r: Recommendation) => ({
    title: r.title, action: r.recommended_action, priority: r.priority, domain: r.domain,
  }));

  const nextActions = recs
    .filter((r) => r.priority === "High")
    .map((r) => firstSentence(r.recommended_action))
    .slice(0, 4);

  const evidence = {
    period: cycle,
    kpis: kpis.map((k) => ({ label: k.label, value: k.value, unit: k.unit, delta: k.delta })),
    projected_next: projected && { value: projected.yhat, unit: moneyUnit },
    top_ranked_item: topItem && { name: topItem.entity, metric: topItem.metric },
    top_skill_premium: topSkill && { skill: topSkill.entity, premium_pct: topSkill.metric },
    best_channel: bestChannel && { channel: bestChannel.entity, metric: bestChannel.metric },
    weak_region: weakRegion && { region: weakRegion.entity, metric: weakRegion.metric },
    recommendations: recs.map((r) => ({ title: r.title, priority: r.priority, domain: r.domain })),
  };

  const brief: ExecutiveBrief = {
    provenance: { engine: "deterministic", provider: providerLabel() },
    cycle, summary, insights, risks, recommendations, nextActions, evidence,
  };

  // --- optional AI enhancement of the narrative ----------------------------
  if (isConfigured()) {
    const enhanced = await enhance(evidence);
    if (enhanced) {
      brief.summary = enhanced.summary || brief.summary;
      if (enhanced.insights?.length) brief.insights = enhanced.insights;
      if (enhanced.risks?.length) brief.risks = enhanced.risks;
      if (enhanced.nextActions?.length) brief.nextActions = enhanced.nextActions;
      brief.provenance.engine = "ai-enhanced";
    }
  }

  return brief;
}

async function enhance(evidence: Record<string, unknown>) {
  const prompt =
    `Here is the structured analytical output for a placement-intelligence cycle. ` +
    `Write an executive brief strictly from this evidence. Respond with ONLY valid JSON of the shape ` +
    `{"summary": string, "insights": string[], "risks": string[], "nextActions": string[]}. ` +
    `Keep summary to 3 sentences, and 3 to 5 items per list.\n\nEVIDENCE:\n${JSON.stringify(evidence, null, 2)}`;
  const raw = await generate({ system: PERSONAS.report_writer.system, prompt, temperature: 0.2 });
  if (!raw) return null;
  try {
    const json = JSON.parse(raw.replace(/^```json\s*/i, "").replace(/```$/, "").trim());
    return json as { summary?: string; insights?: string[]; risks?: string[]; nextActions?: string[] };
  } catch {
    return null;
  }
}

function firstSentence(text: string): string {
  const m = text.match(/^[^.]+\./);
  return (m ? m[0] : text).trim();
}

File: api/athena_api/brief.py
"""Executive Brief builder (Python, for the API and n8n).

Deterministic first: a full board-ready brief is assembled from the result
tables and works with every AI provider off. If a provider is configured, the AI
layer rewrites only the narrative prose. Numbers and recommendations always come
from the deterministic layer.
"""

from __future__ import annotations

import json
import re

from athena.config import MODULE

from . import ai
from .data import forecast, kpis, ranking, recommendations


def _by_key(rows, key):
    return next((r for r in rows if r["metric_key"] == key), None)


def _fmt_delta(d, unit=""):
    if d is None:
        return ""
    return f"{'up' if d >= 0 else 'down'} {abs(d):.1f}{unit}"


def _first_sentence(text: str) -> str:
    m = re.match(r"^[^.]+\.", text)
    return (m.group(0) if m else text).strip()


def build_brief(conn, module=MODULE) -> dict:
    k = kpis(conn, module)
    recs = recommendations(conn, module)
    skill_prem = ranking(conn, "skill_salary_premium", module)
    region = ranking(conn, "region_placement_rate", module)
    channel = ranking(conn, "channel_effectiveness", module)
    ctc_fc = forecast(conn, "median_ctc_cycle", module)

    rate = _by_key(k, "placement_rate")
    ctc = _by_key(k, "median_ctc")
    placed = _by_key(k, "total_placements")
    top_sector = _by_key(k, "top_sector_hiring")
    cycle = (rate["context"].replace("Cycle ", "") if rate and rate["context"] else "latest")
    projected = next((p for p in ctc_fc if p["is_forecast"]), None)
    top_skill = skill_prem[0] if skill_prem else None
    best_channel = channel[0] if channel else None
    weak_region = region[0] if region else None

    summary = (
        f"In the {cycle} cycle, "
        f"{round(placed['value']):,} students were placed at a "
        f"{rate['value']:.1f}% placement rate ({_fmt_delta(rate['delta'], ' points')} on the prior cycle), "
        f"with a median CTC of ₹{ctc['value']:.1f} LPA ({_fmt_delta(ctc['delta'], ' LPA')}). "
        + (f"{top_sector['context']} is the largest recruiting sector. " if top_sector else "")
        + (f"The salary trajectory projects to ₹{projected['yhat']:.1f} LPA next cycle." if projected else "")
    )

    insights = [x for x in [
        top_skill and f"{top_skill['entity']} carries the clearest salary premium at "
                      f"{'+' if top_skill['metric'] >= 0 else ''}{top_skill['metric']:.1f}% over the median.",
        best_channel and f"{best_channel['entity']} is the highest-yield hiring channel at "
                         f"{best_channel['metric']:.1f}%.",
        weak_region and f"{weak_region['entity']} is the weakest region at {weak_region['metric']:.1f}%, "
                        f"below the {rate['value']:.1f}% average.",
    ] if x]

    high = [r for r in recs if r["priority"] == "High"]
    risks = ([r["observation"] for r in high]
             + ([f"Placement is concentrated away from {weak_region['entity']}, "
                 f"leaving that region exposed."] if weak_region else []))[:4]

    recommendations_out = [{
        "title": r["title"], "action": r["recommended_action"],
        "priority": r["priority"], "domain": r["domain"],
    } for r in recs[:5]]

    next_actions = [_first_sentence(r["recommended_action"]) for r in high][:4]

    evidence = {
        "cycle": cycle,
        "placement_rate": rate and rate["value"], "placement_rate_delta": rate and rate["delta"],
        "median_ctc": ctc and ctc["value"], "median_ctc_delta": ctc and ctc["delta"],
        "students_placed": placed and placed["value"],
        "top_sector": top_sector and top_sector["context"],
        "projected_next_ctc": projected and projected["yhat"],
        "top_skill_premium": top_skill and {"skill": top_skill["entity"], "premium_pct": top_skill["metric"]},
        "best_channel": best_channel and {"channel": best_channel["entity"], "rate": best_channel["metric"]},
        "weak_region": weak_region and {"region": weak_region["entity"], "rate": weak_region["metric"]},
        "recommendations": [{"title": r["title"], "priority": r["priority"], "domain": r["domain"]} for r in recs],
    }

    brief = {
        "provenance": {"engine": "deterministic", "provider": ai.provider_label()},
        "cycle": cycle, "summary": summary, "insights": insights, "risks": risks,
        "recommendations": recommendations_out, "next_actions": next_actions, "evidence": evidence,
    }

    if ai.is_configured():
        enhanced = _enhance(evidence)
        if enhanced:
            brief["summary"] = enhanced.get("summary") or brief["summary"]
            if enhanced.get("insights"):
                brief["insights"] = enhanced["insights"]
            if enhanced.get("risks"):
                brief["risks"] = enhanced["risks"]
            if enhanced.get("next_actions"):
                brief["next_actions"] = enhanced["next_actions"]
            brief["provenance"]["engine"] = "ai-enhanced"

    return brief


def _enhance(evidence: dict):
    prompt = (
        "Here is the structured analytical output for a placement-intelligence cycle. "
        "Write an executive brief strictly from this evidence. Respond with ONLY valid JSON of the shape "
        '{"summary": string, "insights": string[], "risks": string[], "next_actions": string[]}. '
        "Keep summary to 3 sentences, 3 to 5 items per list.\n\nEVIDENCE:\n"
        + json.dumps(evidence, indent=2, default=str)
    )
    raw = ai.generate(ai.PERSONAS["report_writer"], prompt, temperature=0.2)
    if not raw:
        return None
    try:
        cleaned = re.sub(r"^```json\s*", "", raw).rstrip("`").strip()
        return json.loads(cleaned)
    except Exception:  # noqa: BLE001
        return None
